### PR TITLE
Add some tests to System.Text.RegularExpressions

### PR DESCRIPTION
**Pull request by hughbe**
- Follows up from the recent cleanup work (#7046)
- Line coverage: 82.8% -> 86%
- Branch coverage: 72.1% -> 76.1%

Just a note, a lot of lines and branches in this lib project won't be hit as its debug code for dumping etc. (that's why I argue CC builds should default to release mode). As such, when I run the tests in release mode I get line and branch coverage of 93.1% and 85% respectively

/cc @priya91

### CODE REVIEW

**Comment by hughbe** [comment]
Test Innerloop Windows_NT Release Build and Test

File: src/System.Text.RegularExpressions/tests/CaptureCollectionTests.cs [hunk 1]
```diff
@@ -45,5 +46,53 @@ public static void GetEnumerator_Invalid()
...
+            CaptureCollection captures = regex.Match("aaabbccccccccccaaaabc").Captures;
+            ICollection collection = captures;
+
+            RegularExpressions.Capture[] copy = new RegularExpressions.Capture[collection.Count + index];
+            collection.CopyTo(copy, index);
+
+            for (int i = index; i < copy.Length; i++)
+            {
+                Assert.Same(captures[i - index], copy[i]);
+            }
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/CaptureCollectionTests.cs]: Could you add a loop:
```C#
for (int i = 0; i < index; i++)
{
    Assert.Null(copy[i]);
}
```
?

File: src/System.Text.RegularExpressions/tests/GroupCollectionTests.cs [hunk 2]
```diff
@@ -2,26 +2,121 @@
...
+            Match matchPos = regex.Match(inputWithGroup);
+            Match matchNeg = regex.Match(inputWithoutGroup);
+
+            //test if Success returns true with group(s) present
+            Assert.True(matchPos.Groups[0].Success);
+
+            //public [] operator calling internal GetGroup()
+            //should return internal _emptygroup thus making
+            //call to Success() false
+           Assert.False(matchNeg.Groups[0].Success);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/GroupCollectionTests.cs]: Nit: indent

File: src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs [hunk 3]
```diff
@@ -91,6 +91,20 @@ public static IEnumerable<object[]> Replace_String_TestData()
             yield return new object[] { @"\d", "0123456789foo4567890foo         ", "#", RegexOptions.RightToLeft, 7, 32, "0123456789foo#######foo         " };
             yield return new object[] { @"\d", "0123456789foo4567890foo         ", "#", RegexOptions.RightToLeft, 0, 32, "0123456789foo4567890foo         " };
             yield return new object[] { @"\d", "0123456789foo4567890foo         ", "#", RegexOptions.RightToLeft, -1, 32, "##########foo#######foo         " };
+
+            yield return new object[] { "([1-9])([1-9])([1-9])def", "abc123def!", "$0", RegexOptions.RightToLeft, -1, 10, "abc123def!" };
+            yield return new object[] { "([1-9])([1-9])([1-9])def", "abc123def!", "$1", RegexOptions.RightToLeft, -1, 10, "abc1!" };
+            yield return new object[] { "([1-9])([1-9])([1-9])def", "abc123def!", "$2", RegexOptions.RightToLeft, -1, 10, "abc2!" };
+            yield return new object[] { "([1-9])([1-9])([1-9])def", "abc123def!", "$3", RegexOptions.RightToLeft, -1, 10, "abc3!" };
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: The difference between these three cases is just whether 1, 2, or 3 is used?  Is there a meaningful difference between them?

**Comment by stephentoub** [comment]
Some nits, otherwise LGTM.

File: src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs (hunk 3 above)
Comment (hughbe) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: I was just going through every single group from 1-3 to make sure it replaces the match with the group

Comment (hughbe) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: ... this doesn't affect coverage, but IMO is a bit more complete

Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: I'm not concerned about code coverage numbers.  How is it more complete?  I'm just trying to understand what the additional value is from having the additional test cases.  For example, what are you imagining changing that would break one of these but not the others?

Comment (hughbe) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: right - my logic behind adding all of the group tests was to make sure that each group number corresponds to the correct group, and there are no off-by-one errors or problems in a situation such as "$X" where `x == groups.Length`

Comment (hughbe) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: But I can remove the $2 and $3 tests if you want

Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: I'm still unclear on the value, but you can leave them.

Comment (hughbe) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: Sorry if I didn't explain it well ;P its late at night here

**Comment by hughbe** [comment]
Thanks for the comments - they've been addressed in a new commit

File: src/System.Text.RegularExpressions/tests/GroupCollectionTests.cs [hunk 4]
```diff
@@ -103,7 +103,14 @@ public void ICollection_CopyTo(int index)
...
             collection.CopyTo(copy, index);
-
+            for (int i = 0; i < index; i++)
+            {
+                Assert.Null(copy[i]);
+            }
+            for (int i = 0; i < index; i++)
+            {
+                Assert.Null(copy[i]);
+            }
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/GroupCollectionTests.cs]: Why is this loop duplicated?

Comment (hughbe) [inline src/System.Text.RegularExpressions/tests/GroupCollectionTests.cs]: whoops, fixed
